### PR TITLE
Added comparable matcher result and applied to shouldContainExactly

### LINKED ISSUE #1851: shouldContainExactly should show diff in IntelliJ
**Please describe the feature you'd like to see including any solutions in mind if you have any**

### PR DESCRIPTION
**Pull request by sksamuel**
Closes #1851

### CODE REVIEW

**Comment by ashishkujoy** [comment]
Looks Good to me.

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt [hunk 1]
```diff
@@ -3,15 +3,35 @@ package io.kotest.matchers.collections
...
+import io.kotest.matchers.should
+import io.kotest.matchers.shouldNot
 import kotlin.jvm.JvmName
 
+/**
+ * Verifies that this collection contains the sub collections provided in the exact given order.
+ *
+ * So, for example, listOf(1,2,3) contains exactly the sub collections:
+ *
+ * [], [1], [2], [3], [1,2], [2,3] and [1,2,3].
```
Comment (Kantis) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt:20]: Suppose this is out of scope for the PR, but I was a bit surprised when reading this comment. I expected `shouldContainExactly` to match if, and only if, the actual collection contains _exactly_(?) the elements in expected. I.e. no more and no less than the given. 

E.g.

```kotlin 
listOf(1, 2, 3, 3) shouldContainExactly listOf(1, 2, 3) -> false
setOf(1, 2, 3, 3) shouldContainExactly listOf(1, 2, 3) -> true 
setOf(1, 2, 3, 3) shouldContainExactly listOf(2, 3) -> false
```

Perhaps it's you might as well be using `x shouldBe y` in that case, but well. 

Just thought I'd throw it out there.

Comment (Kantis) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt:20]: Actually I think this is the actual behaviour, so the comment is lying? 

Given the below test, `actual` contains the subcollection defined in `expected`, but also an additional element causing the test to fail?
bild

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt [hunk 2]
```diff
@@ -45,10 +65,23 @@ fun <T, C : Collection<T>> containExactly(expected: C): Matcher<C?> = neverNullM
       sb.toString()
    }
 
-   MatcherResult(
-      passed,
-      failureMessage
-   ) { "Collection should not be exactly ${expected.printed().value}" }
+   val negatedFailureMessage = { "Collection should not contain exactly ${expected.printed().value}" }
+
+   if (actual.size <= 20 && expected.size <= 20) {
```
Comment (LeoColman) [inline kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt]: Maybe this 20 could be extracted to a system property, just for the sake of configurability.

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt (hunk 1 above)
Comment (Kantis) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt:20]: I realize I got it the wrong way around. The test that caused failure has an extra in actual, not in expected. Sorry 😞

File: kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt (hunk 2 above)
Comment (sksamuel) [reply kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/containExactly.kt]: Yep good idea

**Pull request merged**